util/buildinfo/buildinfo.go

```diff
@@ -2,136 +2,208 @@ package buildinfo
-					Pin:  di,
+			if _, ok := mbs[sourceID.URL]; !ok {
+				mbs[sourceID.URL] = binfotypes.Source{
+					Type: binfotypes.SourceTypeHTTP,
+					Ref:  urlutil.RedactCredentials(sourceID.URL),
+					Pin:  pin,
 				}
 			}
 		}
 	}
 
 	// Leftovers build deps in image config. Mostly duplicated ones we
-	// don't need but there is an edge case if no instruction except source's
-	// one is defined (eg. FROM ...) that can be valid so take it into account.
-	for _, bi := range icbi {
-		if bi.Type != exptypes.BuildInfoTypeDockerImage {
+	// don't need but there is an edge case if no instruction except sources
+	// one is defined (e.g. FROM ...) that can be valid so take it into account.
+	for _, ics := range imageConfigBuildInfo.Sources {
+		if ics.Type != binfotypes.SourceTypeDockerImage {
 			continue
 		}
-		if _, ok := mbis[bi.Alias]; !ok {
-			parsed, err := reference.ParseNormalizedNamed(bi.Ref)
+		if _, ok := mbs[ics.Alias]; !ok {
+			parsed, err := reference.ParseNormalizedNamed(ics.Ref)
 			if err != nil {
-				return nil, errors.Wrapf(err, "failed to parse %s", bi.Ref)
+				return nil, errors.Wrapf(err, "failed to parse %s", ics.Ref)
 			}
-			mbis[bi.Alias] = exptypes.BuildInfo{
-				Type: exptypes.BuildInfoTypeDockerImage,
+			mbs[ics.Alias] = binfotypes.Source{
+				Type: binfotypes.SourceTypeDockerImage,
 				Ref:  reference.TagNameOnly(parsed).String(),
-				Pin:  bi.Pin,
+				Pin:  ics.Pin,
 			}
 		}
 	}
 
-	bis := make([]exptypes.BuildInfo, 0, len(mbis))
-	for _, bi := range mbis {
-		bis = append(bis, bi)
+	srcs := make([]binfotypes.Source, 0, len(mbs))
+	for _, bs := range mbs {
+		srcs = append(srcs, bs)
 	}
-	sort.Slice(bis, func(i, j int) bool {
-		return bis[i].Ref < bis[j].Ref
+	sort.Slice(srcs, func(i, j int) bool {
+		return srcs[i].Ref < srcs[j].Ref
 	})
 
-	return json.Marshal(map[string][]exptypes.BuildInfo{
-		"sources": bis,
-	})
+	return srcs, nil
 }
 
-// imageConfigBuildInfo returns build dependencies from image config
-func imageConfigBuildInfo(imageConfig []byte) ([]exptypes.BuildInfo, error) {
+// FromImageConfig returns build dependencies from image config.
+func FromImageConfig(imageConfig []byte) (bi binfotypes.BuildInfo, _ error) {
 	if len(imageConfig) == 0 {
-		return nil, nil
-	}
-	var config struct {
-		BuildInfo []byte `json:"moby.buildkit.buildinfo.v1,omitempty"`
+		return bi, nil
 	}
+	var config binfotypes.ImageConfig
 	if err := json.Unmarshal(imageConfig, &config); err != nil {
-		return nil, errors.Wrap(err, "failed to unmarshal buildinfo from config")
+		return bi, errors.Wrap(err, "failed to unmarshal buildinfo from image config")
 	}
 	if len(config.BuildInfo) == 0 {
-		return nil, nil
+		return bi, nil
 	}
-	var bi []exptypes.BuildInfo
 	if err := json.Unmarshal(config.BuildInfo, &bi); err != nil {
-		return nil, errors.Wrapf(err, "failed to unmarshal %s", ImageConfigField)
+		return bi, errors.Wrap(err, "failed to unmarshal buildinfo")
 	}
 	return bi, nil
 }
+
+var knownAttrs = []string{
```

Comment:
Is this list Dockerfile specific or generic enough to be in here?

Reply:
Yes this is Dockerfile specific atm: https://github.com/moby/buildkit/blob/ffe2301031c8f8bfb8d5fc5034e5e509c5624913/frontend/dockerfile/builder/build.go#L44-L61